### DISCUSSION TITLE
Apple Sign-In Reauthentication Error

### DISCUSSION THREAD

**Discussion by szaycev** [firebase/firebase-ios-sdk#11016] [Q&A]
Hi all,
I need to add functionality to delete a user from the application. I use firebase and only Apple provider is used for registration and authorization.
When I try to re-authenticate a user using `Auth.auth().currentUser?.reauthenticate(with:)` I always get an error:

`error: Error Domain=FIRAuthErrorDomain Code=17094 "Duplicate credential received. Please try again with a new credential." UserInfo={NSLocalizedDescription=Duplicate credential received. Please try again with a new credential., FIRAuthErrorUserInfoNameKey=ERROR_MISSING_OR_INVALID_NONCE}`

Here is the code I am using to delete a user:

```
func deleteAccountApple() async throws {
    guard let user = Auth.auth().currentUser else {
        throw AccountDeletionError.notSignedIn
    }
    do {
        try await user.delete()
    } catch {
        let nsError = error as NSError
        let authError = AuthErrorCode.Code(rawValue: nsError.code)

        if authError == .requiresRecentLogin {
            do {
                try await reauthenticateFirebaseUser()
                try await user.delete()
            } catch {
                throw AccountDeletionError.failedToDeleteFirebaseAccount(error)
            }
        } else {
            throw AccountDeletionError.failedToDeleteFirebaseAccount(error)
        }
    }
}

private func reauthenticateFirebaseUser() async throws {
    guard let user = Auth.auth().currentUser else {
        throw AccountDeletionError.notSignedIn
    }

    let idToken = try await user.getIDToken()
    let nonce: String = .randomNonceString
  
    let credentials = OAuthProvider.credential(
        withProviderID: "apple.com",
        idToken: idToken,
        rawNonce: nonce
    )

    try await user.reauthenticate(with: credentials)
}
```

I have tried to generate a new `nonce` and also tried to save and use the same `nonce` as when the user was authorized but I always get the error.

How to get valid credentials to re-authenticate a user who is logged in with Apple?

I would appreciate any help!

**Comment by heoseongjin** [comment]
> Hi all, I need to add functionality to delete a user from the application. I use firebase and only Apple provider is used for registration and authorization. When I try to re-authenticate a user using `Auth.auth().currentUser?.reauthenticate(with:)` I always get an error:
> 
> `error: Error Domain=FIRAuthErrorDomain Code=17094 "Duplicate credential received. Please try again with a new credential." UserInfo={NSLocalizedDescription=Duplicate credential received. Please try again with a new credential., FIRAuthErrorUserInfoNameKey=ERROR_MISSING_OR_INVALID_NONCE}`
> 
> Here is the code I am using to delete a user:
> 
> ```
> func deleteAccountApple() async throws {
>     guard let user = Auth.auth().currentUser else {
>         throw AccountDeletionError.notSignedIn
>     }
>     do {
>         try await user.delete()
>     } catch {
>         let nsError = error as NSError
>         let authError = AuthErrorCode.Code(rawValue: nsError.code)
> 
>         if authError == .requiresRecentLogin {
>             do {
>                 try await reauthenticateFirebaseUser()
>                 try await user.delete()
>             } catch {
>                 throw AccountDeletionError.failedToDeleteFirebaseAccount(error)
>             }
>         } else {
>             throw AccountDeletionError.failedToDeleteFirebaseAccount(error)
>         }
>     }
> }
> 
> private func reauthenticateFirebaseUser() async throws {
>     guard let user = Auth.auth().currentUser else {
>         throw AccountDeletionError.notSignedIn
>     }
> 
>     let idToken = try await user.getIDToken()
>     let nonce: String = .randomNonceString
>   
>     let credentials = OAuthProvider.credential(
>         withProviderID: "apple.com",
>         idToken: idToken,
>         rawNonce: nonce
>     )
> 
>     try await user.reauthenticate(with: credentials)
> }
> ```
> 
> I have tried to generate a new `nonce` and also tried to save and use the same `nonce` as when the user was authorized but I always get the error.
> 
> How to get valid credentials to re-authenticate a user who is logged in with Apple?
> 
> I would appreciate any help!

Hi there,

The issue you're facing is likely due to using `OAuthProvider.credential` with an Apple ID token and a new nonce when trying to reauthenticate the user. This is not the correct way to reauthenticate users who have signed in with Apple.

To reauthenticate the user properly, you need to follow the same process you used when the user initially signed in with Apple. You should use the `ASAuthorizationAppleIDProvider` to create an `ASAuthorizationAppleIDRequest` and present the `ASAuthorizationController` again to obtain the user's Apple credential. After that, you can use this credential to reauthenticate the user with Firebase.

By doing so, you will be able to properly reauthenticate your user who is logged in with Apple and then proceed with account deletion.

☺️
